spring-kafka/src/main/java/org/springframework/kafka/listener/FailedRecordProcessor.java

```diff
@@ -0,0 +1,222 @@
+import org.springframework.kafka.support.serializer.DeserializationException;
+import org.springframework.lang.Nullable;
+import org.springframework.messaging.converter.MessageConversionException;
+import org.springframework.messaging.handler.invocation.MethodArgumentResolutionException;
+import org.springframework.util.Assert;
+import org.springframework.util.backoff.BackOff;
+import org.springframework.util.backoff.FixedBackOff;
+
+/**
+ * Common super class for classes that deal with failing to consume a consumer record.
+ *
+ * @author Gary Russell
+ * @since 2.3.1
+ *
+ */
+public abstract class FailedRecordProcessor {
+
+	private static final BiPredicate<ConsumerRecord<?, ?>, Exception> ALWAYS_SKIP_PREDICATE = (r, e) -> true;
+
+	private static final BiPredicate<ConsumerRecord<?, ?>, Exception> NEVER_SKIP_PREDICATE = (r, e) -> false;
+
+	protected static final LogAccessor LOGGER =
+			new LogAccessor(LogFactory.getLog(SeekToCurrentErrorHandler.class)); // NOSONAR
+
+	private final FailedRecordTracker failureTracker;
+
+	private BinaryExceptionClassifier classifier;
+
+	private boolean commitRecovered;
+
+	protected FailedRecordProcessor(@Nullable BiConsumer<ConsumerRecord<?, ?>, Exception> recoverer, BackOff backOff) {
+		this.failureTracker = new FailedRecordTracker(recoverer, backOff, LOGGER);
+		this.classifier = configureDefaultClassifier();
+	}
+
+	/**
+	 * TODO: remove when the deprecated dependent CTORs are removed.
+	 * @param recoverer the recoverer.
+	 * @param maxFailures the max failures.
+	 */
+	FailedRecordProcessor(@Nullable BiConsumer<ConsumerRecord<?, ?>, Exception> recoverer, int maxFailures) {
+		this.failureTracker = new FailedRecordTracker(recoverer, new FixedBackOff(0L, maxFailures - 1), LOGGER);
+		this.classifier = configureDefaultClassifier();
+	}
+
+	/**
+	 * Return the exception classifier.
+	 * @return the classifier.
+	 */
+	protected BinaryExceptionClassifier getClassifier() {
+		return this.classifier;
+	}
+
+	/**
+	 * Set an exception classifier to determine whether the exception should cause a retry
+	 * (until exhaustion) or not. If not, we go straight to the recoverer. By default,
+	 * the following exceptions will not be retried:
+	 * <ul>
+	 * <li>{@link DeserializationException}</li>
+	 * <li>{@link MessageConversionException}</li>
+	 * <li>{@link MethodArgumentResolutionException}</li>
+	 * <li>{@link NoSuchMethodException}</li>
+	 * <li>{@link ClassCastException}</li>
+	 * </ul>
+	 * All others will be retried.
+	 * The classifier's {@link BinaryExceptionClassifier#setTraverseCauses(boolean) traverseCauses}
+	 * will be set to true because the container always wraps exceptions in a
+	 * {@link ListenerExecutionFailedException}.
+	 * This replaces the default classifier.
+	 * @param classifier the classifier.
+	 */
+	public void setClassifier(BinaryExceptionClassifier classifier) {
+		Assert.notNull(classifier, "'classifier' + cannot be null");
+		classifier.setTraverseCauses(true);
```

Comment:
Can't we request `Map<Class<? extends Throwable>, Boolean>` instead here and create a `BinaryExceptionClassifier` internaly to avoid such a mutation of the provided one?